src/components/legend/draw.js

```diff
@@ -77,8 +77,43 @@ function drawOne(gd, opts) {
 
     var legendData;
     if(!inHover) {
-        if(!gd.calcdata) return;
-        legendData = fullLayout.showlegend && getLegendData(gd.calcdata, legendObj, fullLayout._legends.length > 1);
+        var calcdata = (gd.calcdata || []).slice();
+
+        var shapes = fullLayout.shapes;
+        for(var i = 0; i < shapes.length; i++) {
+            var shape = shapes[i];
+            if(!shape.showlegend) continue;
+
+            var shapeLegend = {
+                _fullInput: shape,
+                index: shape._index,
+                name: shape.name || shape.label.text || ('shape ' + shape._index),
```

Comment:
Should there be a way to set text to appear in the legend for a`newshape` and still use `texttemplate` for the `newshape`. I can do it for shapes that aren't `newshape`. For example 

```
fig.add_shape(
    showlegend=True,
    name="Change from peak",
    type="line",
    xref="x",
    x0="2012-09-01",
    y0=16.1,
    x1="2020-04-01",
    y1=5,
    line_width=4,
    label=dict(
        text='abc',
        texttemplate="Change from peak: %{dy:.2f}",
        textposition="top center",
        font=dict(size=10),
    ),
    fillcolor="grey",
)
```

Looks like this:

<img width="752" alt="image" src="https://github.com/plotly/plotly.js/assets/12749831/77329b50-d576-487e-8f68-9ee92bb56c3d">

But for `newshape`, there is no `name`. If you set just the `label.texttemplate`, a `newshape` looks like this in the legend:

```
        newshape=dict(
            showlegend=True,
            label=dict(texttemplate="Change: %{dy:.2f}")
        ),
```
<img width="792" alt="image" src="https://github.com/plotly/plotly.js/assets/12749831/54275991-b4c7-4529-870d-fe6ffad53c73">

And it shows the same with `label.text` specified:
```
        newshape=dict(
            showlegend=True,
            label=dict(
                text="abc",
                texttemplate="Change: %{dy:.2f}"
            )
```
So right now, I need to exclude `label.texttemplate` for it to work:

```
        newshape=dict(
            showlegend=True,
            label=dict(
                text="abc",
#                 texttemplate="Change: %{dy:.2f}"
            )
        ),
```

<img width="788" alt="image" src="https://github.com/plotly/plotly.js/assets/12749831/c88aa98f-b7fb-4b79-baf8-9a1af2ef98a4">

Reply:
Good call. Addressed in [6343338](https://github.com/plotly/plotly.js/pull/6653/commits/634333884bbcef437a3271b0a3a942cf0f0e5c1d).

Reply:
Good catch. Fixed by [948fd02](https://github.com/plotly/plotly.js/pull/6653/commits/948fd02099d62c34b89b3984b2846f3e0cde8565).

Reply:
Works for me now. Thanks!